lib/rack/response.rb

```diff
@@ -29,7 +29,7 @@ class Response
 
     def initialize(body = [], status = 200, header = {})
       @status = status.to_i
-      @header = Utils::HeaderHash.new.merge(header)
+      @header = Utils::HeaderHash.new(header)
```

Comment:
I think this fail when `header` is a `HeaderHash` already. In that case `new` just return the same object what makes this object to mutate the original argument too. I do think it does make sense to avoid the useless instantiation when `header` is a regular hash but we need to take in consideration when it is a `HeaderHash`, probably changing `new` to dup the argument in that case.

Reply:
@rafaelfranca Correct me if I'm wrong. From this line of code, it seems safe to pass both `Utils::HeaderHash` and `Hash`: https://github.com/rack/rack/blob/master/lib/rack/utils.rb#L415

```ruby
def self.new(hash = {})
        HeaderHash === hash ? hash : super(hash)
end
```

Reply:
Yes, it is safe, but the object passed will be mutated.

For example, before this change:

```ruby
header = Rack::Utils::HeaderHash.new

response = Rack::Response.new([], 200, header)
response.header['a'] = 1
p header # => {}
```

With your change:

```ruby
header = Rack::Utils::HeaderHash.new

response = Rack::Response.new([], 200, header)
response.header['a'] = 1
p header # => { "a" => 1 }
```

This only happens when it is a `HeaderHash`:

```
header = {}

response = Rack::Response.new([], 200, header)
response.header['a'] = 1
p header # => {}
```

Reply:
@rafaelfranca Thanks for the pointer.
It only happens with `HeaderHash`, because `Hash` is iterated during `HeaderHash#initialize`.

```ruby
      def initialize(hash = {})
        super()
        @names = {}
        hash.each { |k, v| self[k] = v }
      end
```

---

If I try to _dup_ `HeaderHash` in `.new`, there is a spec that fails:

```ruby
      def self.new(hash = {})
        HeaderHash === hash ? hash.dup : super(hash)
      end
```

```ruby
# test/spec_utils.rb:677

  it "avoid unnecessary object creation if possible" do
    a = Rack::Utils::HeaderHash.new("foo" => "bar")
    b = Rack::Utils::HeaderHash.new(a)
    b.object_id.must_equal a.object_id
    b.must_equal a
  end
```

This makes sense for performance POV, but doesn't guarantee immutability.

So I grep'd for the usage of `HeaderHash` and discovered it's used only with `new` (**without** `merge`) in the rest of the code base. These are middleware, so I'm not sure if we care about immutability over there.

---

A solution to preserve the current behavior of `.new` and to make `Response#initialize` faster for the most common case (`header` is a `Hash`), we can introduce `HeaderHash.build`, as a special case:

```ruby
      def self.build(hash)
        new(HeaderHash === hash ? hash.dup : hash)
      end
```

This will make **existing** specs to pass, and also a **new** one that I wrote:

```ruby
# test/spec_response.rb
  it "doesn't mutate given headers" do
    [{}, Rack::Utils::HeaderHash.new].each do |header|
      response = Rack::Response.new([], 200, header)
      response.header["Content-Type"] = "text/plain"
      response.header["Content-Type"].must_equal "text/plain"

      header.wont_include("Content-Type")
    end
  end
```

---

Here's the results before and after this implementation:

```shell
$ wrk -t 2 http://localhost:9292/
Running 10s test @ http://localhost:9292/
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.00ms  480.28us  13.76ms   86.93%
    Req/Sec     5.13k   237.42     5.39k    91.09%
  103020 requests in 10.10s, 3.93MB read
Requests/sec:  10200.58
Transfer/sec:    398.47KB
```

```shell
$ wrk -t 2 http://localhost:9292/
Running 10s test @ http://localhost:9292/
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     0.96ms  453.09us  12.87ms   87.46%
    Req/Sec     5.35k   269.14     5.69k    90.50%
  106459 requests in 10.00s, 4.06MB read
Requests/sec:  10644.92
Transfer/sec:    415.83KB
```

---

Please let me know what you think. Thanks! 👍

Reply:
@tenderlove what do you think about the current behavior of `new`? Should we care about immutability of the `HeaderHash` in the `Response#initialize` if we don't care in other places?

Reply:
I don't think we should care about immutability.